sdk-lib/src/main/java/cash/z/ecc/android/sdk/tool/WalletBirthdayTool.kt

```diff
@@ -76,44 +97,75 @@ class WalletBirthdayTool(appContext: Context) {
+
+            twig("found ${treeFiles.size} sapling tree checkpoints: $treeFiles")
+
+            return getFirstValidWalletBirthday(context, directory, treeFiles)
+        }
+
+        private fun getFilteredFileNames(
+            context: Context,
+            directory: String,
+            birthdayHeight: Int? = null,
+        ): List<String> {
+            val unfilteredTreeFiles = listBirthdayDirectoryContents(context, directory)
+            if (unfilteredTreeFiles.isNullOrEmpty()) {
+                throw BirthdayException.MissingBirthdayFilesException(directory)
+            }
+
+            val filteredTreeFiles = unfilteredTreeFiles
+                .sortDescending()
+                .filter { filename ->
+                    birthdayHeight?.let { birthdayHeight(filename) <= it } ?: true
                 }
-            } catch (t: Throwable) {
+
+            if (filteredTreeFiles.isEmpty()) {
                 throw BirthdayException.BirthdayFileNotFoundException(
                     directory,
                     birthdayHeight
                 )
             }
-            try {
-                val reader = JsonReader(
-                    InputStreamReader(context.assets.open("$directory/$file"))
-                )
-                return Gson().fromJson(reader, WalletBirthday::class.java)
-            } catch (t: Throwable) {
-                throw BirthdayException.MalformattedBirthdayFilesException(
-                    directory,
-                    treeFiles[0]
-                )
+
+            return filteredTreeFiles
+        }
+
+        /**
+         * @param treeFiles A list of files, sorted in descending order based on `int` value of the first part of the filename.
+         */
+        @VisibleForTesting
+        internal fun getFirstValidWalletBirthday(
+            context: Context,
+            directory: String,
+            treeFiles: List<String>
+        ): WalletBirthday {
+            var lastException: Exception? = null
+            treeFiles.forEach { treefile ->
+                try {
+                    context.assets.open("$directory/$treefile").use { inputStream ->
+                        InputStreamReader(inputStream).use { inputStreamReader ->
+                            JsonReader(inputStreamReader).use { jsonReader ->
+                                return Gson().fromJson(jsonReader, WalletBirthday::class.java)
+                            }
+                        }
+                    }
+                } catch (t: Throwable) {
+                    val exception = BirthdayException.MalformattedBirthdayFilesException(
+                        directory,
+                        treefile
+                    )
+                    lastException = exception
+
+                    if (IS_FALLBACK_ON_FAILURE) {
+                        // TODO: If we ever add crash analytics hooks, this would be something to report
```

Comment:
The SDK should report exceptions and never purposely crash, clients will handle them as they see fit. Some would choose to use crash reporting, others won't.

Reply:
Agreed.  Currently we don't have a mechanism for injecting crash analytics from the consumer of the SDK, so that's probably a new feature to consider for the SDK itself.